### ISSUE TITLE
Rate limit reduced then raised within seconds

### ISSUE THREAD

**Issue by bboreham** [fluxcd/flux#2891]
**Describe the bug**

In fluxd logs, rate limit to one site is reported being reduced then raised within seconds.

**To Reproduce**

Unclear, but several Flux maintainers have access to the cluster - it's Weave Cloud staging.

**Expected behavior**

After reducing the rate limit, leave it down for at least a few minutes, maybe an hour or more.

**Logs**

I wasn't sure if the whole logs might contain private information so filtered to just lines mentioning rate limiting.

```
ts=2020-03-02T12:34:07.462645791Z caller=rate_limiter.go:71 component=ratelimiter info="reducing rate limit" host=quay.io limit=5.00
ts=2020-03-02T12:34:07.463004422Z caller=warming.go:180 component=warmer canonical_name=quay.io/uswitch/kiam auth={map[]} err="requesting tags: error parsing HTTP 429 response body: invalid character '<' looking for beginning of value: \"<html>\\r\\n<head><title>429 Too Many Requests</title></head>\\r\\n<body bgcolor=\\\"white\\\">\\r\\n<center><h1>429 Too Many Requests</h1></center>\\r\\n<hr><center>nginx/1.12.1</center>\\r\\n</body>\\r\\n</html>\\r\\n\""
ts=2020-03-02T12:34:10.203194851Z caller=rate_limiter.go:88 component=ratelimiter info="increasing rate limit" host=quay.io limit=7.50
ts=2020-03-02T12:34:10.642936171Z caller=rate_limiter.go:88 component=ratelimiter info="increasing rate limit" host=quay.io limit=10.00
...
ts=2020-03-02T12:57:11.48896275Z caller=rate_limiter.go:71 component=ratelimiter info="reducing rate limit" host=quay.io limit=5.00
ts=2020-03-02T12:57:11.489245675Z caller=warming.go:180 component=warmer canonical_name=quay.io/damien.lespiau/quay-exporter auth={map[]} err="requesting tags: error parsing HTTP 429 response body: invalid character '<' looking for beginning of value: \"<html>\\r\\n<head><title>429 Too Many Requests</title></head>\\r\\n<body bgcolor=\\\"white\\\">\\r\\n<center><h1>429 Too Many Requests</h1></center>\\r\\n<hr><center>nginx/1.12.1</center>\\r\\n</body>\\r\\n</html>\\r\\n\""
ts=2020-03-02T12:57:14.014239938Z caller=rate_limiter.go:88 component=ratelimiter info="increasing rate limit" host=quay.io limit=7.50
ts=2020-03-02T12:58:02.559620961Z caller=rate_limiter.go:88 component=ratelimiter info="increasing rate limit" host=quay.io limit=10.00
...
ts=2020-03-02T13:05:12.036048871Z caller=rate_limiter.go:71 component=ratelimiter info="reducing rate limit" host=quay.io limit=5.00
ts=2020-03-02T13:05:12.036205679Z caller=warming.go:180 component=warmer canonical_name=quay.io/cortexproject/cortex auth={map[]} err="requesting tags: error parsing HTTP 429 response body: invalid character '<' looking for beginning of value: \"<html>\\r\\n<head><title>429 Too Many Requests</title></head>\\r\\n<body bgcolor=\\\"white\\\">\\r\\n<center><h1>429 Too Many Requests</h1></center>\\r\\n<hr><center>nginx/1.12.1</center>\\r\\n</body>\\r\\n</html>\\r\\n\""
ts=2020-03-02T13:05:15.498842642Z caller=rate_limiter.go:88 component=ratelimiter info="increasing rate limit" host=quay.io limit=7.50
```

**Additional context**

- Flux version: `master-6b481a4d`
- Kubernetes version: 1.14.9
- Git provider: github
- Container registry provider: quay.io (and others)

**Comment by bboreham** [comment]
Another example supplied on Slack today:
```
ts=2020-03-18T05:53:07.694453704Z caller=images.go:17 component=sync-loop msg="polling for new images for automated workloads"
ts=2020-03-18T05:53:09.065221797Z caller=rate_limiter.go:71 component=ratelimiter info="reducing rate limit" host=quay.io limit=25.00
ts=2020-03-18T05:53:09.06538333Z caller=warming.go:180 component=warmer canonical_name=quay.io/uswitch/kiam auth="{map[<acc no>.dkr.ecr.us-east-1.amazonaws.com:<registry creds for AWS@<acc no>.dkr.ecr.us-east-1.amazonaws.com, from /root/.docker/config.json>]}" err="requesting tags: error parsing HTTP 429 response body: invalid character '<' looking for beginning of value: \"<html>\\r\\n<head><title>429 Too Many Requests</title></head>\\r\\n<body bgcolor=\\\"white\\\">\\r\\n<center><h1>429 Too Many Requests</h1></center>\\r\\n<hr><center>nginx/1.12.1</center>\\r\\n</body>\\r\\n</html>\\r\\n\""
ts=2020-03-18T05:53:09.471164475Z caller=rate_limiter.go:88 component=ratelimiter info="increasing rate limit" host=quay.io limit=37.50
```

**Comment by ashlineldridge** [comment]
I'm currently seeing this as well after installing cert-manager v0.15.0.

```
{"auth":"{map[]}","caller":"warming.go:180","canonical_name":"quay.io/jetstack/cert-manager-webhook","component":"warmer","err":"requesting tags: error parsing HTTP 429 response body: invalid character '\u003c' looking for beginning of value: \"\u003chtml\u003e\\r\\n\u003chead\u003e\u003ctitle\u003e429 Too Many Requests\u003c/title\u003e\u003c/head\u003e\\r\\n\u003cbody bgcolor=\\\"white\\\"\u003e\\r\\n\u003ccenter\u003e\u003ch1\u003e429 Too Many Requests\u003c/h1\u003e\u003c/center\u003e\\r\\n\u003chr\u003e\u003ccenter\u003enginx/1.12.1\u003c/center\u003e\\r\\n\u003c/body\u003e\\r\\n\u003c/html\u003e\\r\\n\"","ts":"2020-05-12T08:58:22.802168912Z"}
{"caller":"rate_limiter.go:88","component":"ratelimiter","host":"quay.io","info":"increasing rate limit","limit":"150.00","ts":"2020-05-12T08:59:19.9442362Z"}
{"caller":"rate_limiter.go:88","component":"ratelimiter","host":"quay.io","info":"increasing rate limit","limit":"200.00","ts":"2020-05-12T08:59:21.398330068Z"}
{"caller":"rate_limiter.go:71","component":"ratelimiter","host":"quay.io","info":"reducing rate limit","limit":"100.00","ts":"2020-05-12T09:12:22.202668552Z"}
{"auth":"{map[]}","caller":"warming.go:180","canonical_name":"quay.io/jetstack/cert-manager-controller","component":"warmer","err":"requesting tags: error parsing HTTP 429 response body: invalid character '\u003c' looking for beginning of value: \"\u003chtml\u003e\\r\\n\u003chead\u003e\u003ctitle\u003e429 Too Many Requests\u003c/title\u003e\u003c/head\u003e\\r\\n\u003cbody bgcolor=\\\"white\\\"\u003e\\r\\n\u003ccenter\u003e\u003ch1\u003e429 Too Many Requests\u003c/h1\u003e\u003c/center\u003e\\r\\n\u003chr\u003e\u003ccenter\u003enginx/1.12.1\u003c/center\u003e\\r\\n\u003c/body\u003e\\r\\n\u003c/html\u003e\\r\\n\"","ts":"2020-05-12T09:12:22.202860845Z"}
{"caller":"rate_limiter.go:88","component":"ratelimiter","host":"quay.io","info":"increasing rate limit","limit":"150.00","ts":"2020-05-12T09:13:20.47999217Z"}
{"caller":"rate_limiter.go:88","component":"ratelimiter","host":"quay.io","info":"increasing rate limit","limit":"200.00","ts":"2020-05-12T09:13:22.321142329Z"}
{"caller":"rate_limiter.go:71","component":"ratelimiter","host":"quay.io","info":"reducing rate limit","limit":"100.00","ts":"2020-05-12T09:37:19.440507985Z"}
{"auth":"{map[]}","caller":"warming.go:180","canonical_name":"quay.io/jetstack/cert-manager-webhook","component":"warmer","err":"requesting tags: error parsing HTTP 429 response body: invalid character '\u003c' looking for beginning of value: \"\u003chtml\u003e\\r\\n\u003chead\u003e\u003ctitle\u003e429 Too Many Requests\u003c/title\u003e\u003c/head\u003e\\r\\n\u003cbody bgcolor=\\\"white\\\"\u003e\\r\\n\u003ccenter\u003e\u003ch1\u003e429 Too Many Requests\u003c/h1\u003e\u003c/center\u003e\\r\\n\u003chr\u003e\u003ccenter\u003enginx/1.12.1\u003c/center\u003e\\r\\n\u003c/body\u003e\\r\\n\u003c/html\u003e\\r\\n\"","ts":"2020-05-12T09:37:19.443242952Z"}
{"caller":"rate_limiter.go:88","component":"ratelimiter","host":"quay.io","info":"increasing rate limit","limit":"150.00","ts":"2020-05-12T09:37:20.621350723Z"}
{"caller":"rate_limiter.go:88","component":"ratelimiter","host":"quay.io","info":"increasing rate limit","limit":"200.00","ts":"2020-05-12T09:37:23.397927728Z"}
{"caller":"warming.go:198","component":"warmer","image":"quay.io/jetstack/cert-manager-webhook","info":"refreshing image","of_which_missing":0,"of_which_refresh":505,"tag_count":505,"to_update":505,"ts":"2020-05-12T09:50:23.203104695Z"}
{"attempted":505,"caller":"warming.go:206","component":"warmer","successful":505,"ts":"2020-05-12T09:50:27.512764322Z","updated":"quay.io/jetstack/cert-manager-webhook"}
{"caller":"warming.go:198","component":"warmer","image":"quay.io/jetstack/cert-manager-controller","info":"refreshing image","of_which_missing":0,"of_which_refresh":675,"tag_count":675,"to_update":675,"ts":"2020-05-12T09:51:20.481462904Z"}
{"attempted":675,"caller":"warming.go:206","component":"warmer","successful":675,"ts":"2020-05-12T09:51:26.006242772Z","updated":"quay.io/jetstack/cert-manager-controller"}
{"caller":"warming.go:198","component":"warmer","image":"quay.io/jetstack/cert-manager-cainjector","info":"refreshing image","of_which_missing":0,"of_which_refresh":357,"tag_count":357,"to_update":357,"ts":"2020-05-12T09:52:22.863919694Z"}
{"attempted":357,"caller":"warming.go:206","component":"warmer","successful":357,"ts":"2020-05-12T09:52:27.108256712Z","updated":"quay.io/jetstack/cert-manager-cainjector"}
{"caller":"rate_limiter.go:71","component":"ratelimiter","host":"quay.io","info":"reducing rate limit","limit":"100.00","ts":"2020-05-12T09:55:23.525117011Z"}
{"auth":"{map[]}","caller":"warming.go:180","canonical_name":"quay.io/jetstack/cert-manager-webhook","component":"warmer","err":"requesting tags: error parsing HTTP 429 response body: invalid character '\u003c' looking for beginning of value: \"\u003chtml\u003e\\r\\n\u003chead\u003e\u003ctitle\u003e429 Too Many Requests\u003c/title\u003e\u003c/head\u003e\\r\\n\u003cbody bgcolor=\\\"white\\\"\u003e\\r\\n\u003ccenter\u003e\u003ch1\u003e429 Too Many Requests\u003c/h1\u003e\u003c/center\u003e\\r\\n\u003chr\u003e\u003ccenter\u003enginx/1.12.1\u003c/center\u003e\\r\\n\u003c/body\u003e\\r\\n\u003c/html\u003e\\r\\n\"","ts":"2020-05-12T09:55:23.525386917Z"}
{"caller":"rate_limiter.go:88","component":"ratelimiter","host":"quay.io","info":"increasing rate limit","limit":"150.00","ts":"2020-05-12T09:56:19.528930444Z"}
{"caller":"rate_limiter.go:88","component":"ratelimiter","host":"quay.io","info":"increasing rate limit","limit":"200.00","ts":"2020-05-12T09:56:21.058342667Z"}
{"caller":"rate_limiter.go:71","component":"ratelimiter","host":"quay.io","info":"reducing rate limit","limit":"100.00","ts":"2020-05-12T09:59:19.575005703Z"}
{"auth":"{map[]}","caller":"warming.go:180","canonical_name":"quay.io/jetstack/cert-manager-controller","component":"warmer","err":"requesting tags: error parsing HTTP 429 response body: invalid character '\u003c' looking for beginning of value: \"\u003chtml\u003e\\r\\n\u003chead\u003e\u003ctitle\u003e429 Too Many Requests\u003c/title\u003e\u003c/head\u003e\\r\\n\u003cbody bgcolor=\\\"white\\\"\u003e\\r\\n\u003ccenter\u003e\u003ch1\u003e429 Too Many Requests\u003c/h1\u003e\u003c/center\u003e\\r\\n\u003chr\u003e\u003ccenter\u003enginx/1.12.1\u003c/center\u003e\\r\\n\u003c/body\u003e\\r\\n\u003c/html\u003e\\r\\n\"","ts":"2020-05-12T09:59:19.575211369Z"}
{"caller":"rate_limiter.go:88","component":"ratelimiter","host":"quay.io","info":"increasing rate limit","limit":"150.00","ts":"2020-05-12T09:59:22.472404434Z"}
{"caller":"rate_limiter.go:88","component":"ratelimiter","host":"quay.io","info":"increasing rate limit","limit":"200.00","ts":"2020-05-12T10:00:19.004120429Z"}
{"caller":"rate_limiter.go:71","component":"ratelimiter","host":"quay.io","info":"reducing rate limit","limit":"100.00","ts":"2020-05-12T10:08:23.213618332Z"}
{"auth":"{map[]}","caller":"warming.go:180","canonical_name":"quay.io/jetstack/cert-manager-webhook","component":"warmer","err":"requesting tags: error parsing HTTP 429 response body: invalid character '\u003c' looking for beginning of value: \"\u003chtml\u003e\\r\\n\u003chead\u003e\u003ctitle\u003e429 Too Many Requests\u003c/title\u003e\u003c/head\u003e\\r\\n\u003cbody bgcolor=\\\"white\\\"\u003e\\r\\n\u003ccenter\u003e\u003ch1\u003e429 Too Many Requests\u003c/h1\u003e\u003c/center\u003e\\r\\n\u003chr\u003e\u003ccenter\u003enginx/1.12.1\u003c/center\u003e\\r\\n\u003c/body\u003e\\r\\n\u003c/html\u003e\\r\\n\"","ts":"2020-05-12T10:08:23.213823981Z"}
```

- Flux version: 1.19.0 (via flux-1.3.0 Helm chart)
- Kubernetes version: 1.16
- Git provider: GitHub
- Container registry: quay.io

**Comment by swade1987** [comment]
We are seeing that too using ...

- Flux version: 1.19.0
- Kubernetes version: 1.7
- Git provider: github.com
- Container registry: quay.io

<img width="976" alt="Screenshot 2020-05-13 at 12 49 37" src="https://user-images.githubusercontent.com/594451/81809062-9d5b9500-9518-11ea-88a4-c21e656ed4f6.png">

**Comment by kingdonb** [comment]
#2986 was merged for 1.20.0, and all the reports for this issue are on 1.19.x or lower, so I assume this can be closed.

If anyone disagrees and can bring a new report, I'll be happy to reopen and take another look. Closing.
